src/analyser.cpp

```diff
@@ -0,0 +1,1509 @@
+            } else if (internalVariable->mType == AnalyserInternalVariable::Type::ALGEBRAIC) {
+                type = AnalyserVariable::Type::ALGEBRAIC;
+            } else {
+                // This is the variable of integration, so skip it.
+
+                continue;
+            }
+
+            auto stateOrVariable = std::shared_ptr<AnalyserVariable> {new AnalyserVariable {}};
+            auto equation = equationMappings[internalVariable->mEquation.lock()];
+
+            stateOrVariable->mPimpl->populate(type, internalVariable->mIndex,
+                                              internalVariable->mInitialisingVariable,
+                                              internalVariable->mVariable,
+                                              equation);
+
+            if (equation != nullptr) {
+                variableMappings[equation] = stateOrVariable;
+            }
+
+            if (type == AnalyserVariable::Type::STATE) {
+                mModel->mPimpl->mStates.push_back(stateOrVariable);
+            } else {
+                mModel->mPimpl->mVariables.push_back(stateOrVariable);
+            }
+        }
+
+        for (const auto &internalEquation : mInternalEquations) {
+            AnalyserEquation::Type type;
+
+            if (internalEquation->mType == AnalyserInternalEquation::Type::TRUE_CONSTANT) {
+                type = AnalyserEquation::Type::TRUE_CONSTANT;
+            } else if (internalEquation->mType == AnalyserInternalEquation::Type::VARIABLE_BASED_CONSTANT) {
+                type = AnalyserEquation::Type::VARIABLE_BASED_CONSTANT;
+            } else if (internalEquation->mType == AnalyserInternalEquation::Type::RATE) {
+                type = AnalyserEquation::Type::RATE;
+            } else {
+                type = AnalyserEquation::Type::ALGEBRAIC;
+            }
+
+            std::vector<AnalyserEquationPtr> dependencies;
+
+            for (const auto &dependency : internalEquation->mDependencies) {
+                dependencies.push_back(equationMappings[dependency]);
+            }
+
+            auto equation = equationMappings[internalEquation];
+
+            equation->mPimpl->populate(type,
+                                       internalEquation->mAst, dependencies,
+                                       internalEquation->mIsStateRateBased,
+                                       variableMappings[equation]);
+
+            mModel->mPimpl->mEquations.push_back(equation);
+        }
+    }
+}
+
+Analyser::Analyser()
+    : mPimpl(new AnalyserImpl {this})
+{
+}
+
+Analyser::~Analyser()
+{
+    delete mPimpl;
+}
+
+AnalyserPtr Analyser::create() noexcept
+{
+    return std::shared_ptr<Analyser> {new Analyser {}};
+}
+
+void Analyser::processModel(const ModelPtr &model)
+{
+    // Make sure that the model is valid before processing it.
+
+    auto validator = Validator::create();
+
+    validator->validateModel(model);
+
+    if (validator->issueCount() > 0) {
```

Comment:
If you make this an `errorCount` then you will avoid warnings? I think that the only warnings at present are around mismatched-but-compatible units, which you deal with anyway?

Reply:
also below

Reply:
I thought the validator only returned errors? So, to use `issueCount()` or `errorCount()` shouldn't make any difference here? Where do we generate warnings about mismatched-but-compatible units? The only warnings I can see are generated by the parser and none of them is relateved mismatched-but-compatible units.

Reply:
Ahh ... It's probably on the list of soon-to-report stuff in some PR which has been waiting for everything else to go in!  I just figured it made your code a little more future-proof, since you'll only ever care about errors (and then only some of them)?

Reply:
My understanding is that the validator can only generate errors and should there be errors/issues the we can't carry on with the analysis. Simple as that.

Reply:
I think it's better to be future-proof and use the counts as they're intended. If at any stage the Validator returns non-error issues this will break ... that's a bit too flimsy for my liking!